### ISSUE TITLE
[no-shadow] Enums can false-positive ESLint's no-shadow rule

### ISSUE THREAD

**Issue by epmatsw** [typescript-eslint/typescript-eslint#325]
**Repro**

```JSON
{
  "rules": {
    "no-shadow": 2
  }
}
```

```TS
const test = 'test';

export const enum MyEnum {
    test = 42
}
```

**Expected Result**

No errors.

**Actual Result**
The `test` enum item gets flagged by `no-shadow`.

**Additional Info**

Seems like maybe we needs an `@typescript-eslint/no-shadow` rule to handle TypeScript's syntax for enums?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.4.2` |
| `@typescript-eslint/parser`        | `1.4.2` |
| `TypeScript`                       | `3.3.3333` |
| `ESLint`                           | `5.14.1` |
| `node`                             | `10.15.0` |
| `npm`                              | `6.7.0` |

**Comment by epmatsw** [comment]
I'm not sure if this should be the parser (for putting an enum item into scope?) or the plugin (it should be in scope, but we need a new TS-specific `no-shadow` to handle it).

**Comment by bradzacher** [comment]
we're working on implementing typescript-compliant scope analysis.
It's hard to do, but better than re-implementing all of the base eslint rules that rely on scope.

**Comment by mysticatea** [comment]
Thank you for this report.

However, this is intentional behavior because the `enum` member `test` literally shadows the upper `test`.

```ts
const f = 777;
enum E1 {
    f = 1, // Shadows the upper `f`.
    g = f, // = 1
}
enum E2 {
    f1 = 1,
    g = f, // This is the upper `f`.
}
```

[Open playground](https://www.typescriptlang.org/play/index.html#src=const%20f%20%3D%20777%3B%0D%0Aenum%20E1%20%7B%0D%0A%20%20%20%20f%20%3D%201%2C%20%2F%2F%20Shadows%20the%20upper%20%60f%60.%0D%0A%20%20%20%20g%20%3D%20f%2C%20%2F%2F%20%3D%201%0D%0A%7D%0D%0Aenum%20E2%20%7B%0D%0A%20%20%20%20f1%20%3D%201%2C%0D%0A%20%20%20%20g%20%3D%20f%2C%20%2F%2F%20This%20is%20the%20upper%20%60f%60.%0D%0A%7D%0D%0A)

**Comment by oigewan** [comment]
Hi @mysticatea.  I was just searching the repo due to the same issue, and I must say that I disagree that this should be intended behavior.  enums are not creating a new scope, they compile down to objects, and developers use them in their code in that fashion.  If I re-write your example using objects instead of enums, it would look like this:
```ts
const f = 777;
const E1 = {
    f: 1,
    g: f,
}
const E2 = {
    f1: 1,
    g: f,
}
```
eslint rightfully does not error on this since the property must be accessed like so: `E1.f`.  This removes the ambiguity that the no-shadow rule wants to prevent.  Given that enum values must be accessed in the exact same way, I think the no-shadow rule should behave in the same way.

Can you elaborate more on why you feel that the no-shadow rule should error out on enum?

**Comment by bradzacher** [comment]
In typescript, you can think of an enum declaration as the following:
A closure which exports all variable declarations as properties on an object it returns.

It's not 100%, but it makes sense if you look at an example [playground](https://www.typescriptlang.org/play/#src=const%20imOutside%20%3D%202%3B%0D%0Aconst%20b%20%3D%202%3B%0D%0Aenum%20Foo%20%7B%0D%0A%20%20%20%20outer%20%3D%20imOutside%2C%0D%0A%20%20%20%20a%20%3D%201%2C%0D%0A%20%20%20%20b%20%3D%20a%2C%0D%0A%20%20%20%20c%20%3D%20b%2C%0D%0A%20%20%20%20%2F%2F%20does%20c%20%3D%3D%20Foo.b%20%3D%3D%20Foo.c%20%3D%3D%201%3F%0D%0A%20%20%20%20%2F%2F%20or%20does%20c%20%3D%3D%20b%20%3D%3D%202%3F%0D%0A%7D)
```ts
const imOutside = 2;
const b = 2;
enum Foo {
    outer = imOutside,
    a = 1,
    b = a,
    c = b,
    // does c == Foo.b == Foo.c == 1?
    // or does c == b == 2?
}
```

In the above example, what does `Foo.c` equal? If you think of an enum as an object like you stated - then your assumption would be that `Foo.c === 2`.
HOWEVER, if you instead think of the enum declaration as a closure, you can tell that the `b = a` declaration ***shadows*** the outer `const b = 2` declaration, which means that in fact `Foo.c === 1`.

This is why the example you provided is not correct. If I rewrite my example using objects:
```ts
const imOutside = 2;
const b = 2;
const Foo = {
    outer: imOutside,
    a: 1,
    b: a,
    c: b,
    // does c == Foo.b == Foo.c == 1?
    // or does c == b == 2?
}
```
Then I can ask the same question - what is the value of `Foo.c`? Because an object expression does not create its own scope, the answer is `Foo.c === 2` (and actually typescript will throw a compile time error at you, because the variable `a` is not defined).

I hope this clarifies it for you. To reiterate:
- `no-shadow` should error out on an enum because enum properties act like variables in a new scope, so they can indeed shadow a variable in the parent scope.
- `no-shadow` should NOT error out on an object because object properties DO NOT act like variables in a new scope, so they CANNOT shadow a variable in the parent scope.

**Comment by oigewan** [comment]
@bradzacher sorry, tried to delete my comment before someone spent time on a response once I realized what was going on.  I did eventually see that enums create a new scope.  That's quite....unexpected.  Thanks for taking the time, and sorry for the bother.

**Comment by mcmar** [comment]
@bradzacher Do you think it's reasonable to request an option for `no-shadow` to allow enums? Seems like a reasonable compromise. Right now, I'm just disabling this rule around my enums when I need to. I'll open a feature request, but not sure if I should do it here or in the core eslint library.

**Comment by bradzacher** [comment]
It's not _unreasonable_, no.
It would have to be in this repo, as the core lib has no concept of enums yet.
It will require the creation of an extension rule in this repo. All around it is a pretty trivial amount of work.

However, for me as an engineer; I would take this as a code smell.
If you are running into this around most enums (or rather enough enums to cause you to want to create an option to work around it), then you might need to either re-evaluate your variable naming patterns, or re-evaluate where you declare your enums (perhaps move them to a new file?).

Creating an option would create a false sense of security, and will potentially cause bugs. With that option, all it takes is one shadow usage by an engineer who doesn't know that enums have a scope, and your enum will have the wrong value.

**Comment by oigewan** [comment]
@bradzacher I agree that as-is, this is code smell since the enum keyword simply creates a scope for assignment.  However, I have an idea of how to alleviate the issue.  Given that you understand the language better than I do, I'm wondering if you have an opinion.

I'm thinking of writing a separate rule in the next month or two (day job is busy) that will require the values of an enum to be primitives.  I would consider basing an enum's value off of a variable instead of the other way around to be a larger bit of code smell than having the property of an object use a shadowed name.  In other words:
```js
// Bad
const someString = "something";
enum Stuff {
  SomeString = someString,
}

// Good
enum Stuff {
  SomeString = "something",
}
const someString = Stuff.SomeString;
```
If paired with the rule I'm wanting to write, I don't think that allowing shadowed enum keys is code smell any more.  Thoughts?

**Comment by bradzacher** [comment]
I would agree with you and that rule would be a good addition.
Flipping the problem on its head gives you safety when turning off `no-shadow` for enums.

**Comment by kg-currenxie** [comment]
I can understand that shadowing happens when you use something from the outer scope as the value, not not as the _key_.

Simplified example:
```
import { white } from './variables'

enum Colors {
  white = '#fff',
}
```

![Screenshot 2020-01-13 at 13 23 44](https://user-images.githubusercontent.com/48229166/72234419-f0274d80-3607-11ea-9a52-5295ff7ad52d.png)

Super annoying :)

**Comment by Klemensas** [comment]
I've just hit @kg-currenxie described case, indeed it's annoying and will force me to disable the rule :(

@oigewan did you had a chance to work on the rule you've proposed? It seems like an adequate solution.

**Comment by bradzacher** [comment]
OOC how often are you hitting it that it's annoying enough to disable the rule entirely?

As mentioned in previous comments, to me that seems like a code smell that there are that many enum member name collisions.

**Comment by Klemensas** [comment]
I guess I should've phrased that differently - I meant that I won't be enabling the rule in the 1st place.

I wanted to introduce the rule into an existing react codebase of medium size. But after learning that I have to work around enum keys to have it play nice it pushes the rule over the limit of being more trouble than use.
The clashes happen mostly with imports, for example, an enum has a key of `Input` and there's a component named `Input` being imported.

IMHO, thinking about my case, the smell arguments aren't very strong. I'm not planning on using any variables for enums, I'd actually like to have the proposed by @oigewan. So I can't think of a good reason why enum member names should be all that different.
As for moving them to a different file, well that's not a good option in my case as well. The components contain types, code, and enums in one file. For me, this structure is tidy enough and works fine.

**Comment by oigewan** [comment]
> I've just hit @kg-currenxie described case, indeed it's annoying and will force me to disable the rule :(
> 
> @oigewan did you had a chance to work on the rule you've proposed? It seems like an adequate solution.

I haven't as I've switched teams and otherwise been quite busy.  However, I am thinking of trying to get started this weekend.

**Comment by oigewan** [comment]
@Klemensas and @bradzacher I've created a pull request with the rule described in this thread:  https://github.com/typescript-eslint/typescript-eslint/pull/1898

If it gets accepted, I can do a follow-up PR which will add an option to the no-shadow rule which will ignore enum members with documentation recommending that the new rule is enabled if the no-shadow option ignoring enum members is enabled.

**Comment by JrSchild** [comment]
Sorry to chip in here. It makes sense that for variables there can be no double declaration. However for types and interfaces that should not be the case as they are not variables, this behavior seems correct if the type and enum is defined in the same file, however once they are split up in separate files this does not work anymore.

E.g. The following does not throw a no-shadow error:
file1.ts
```typescript
export interface Car {
  color: string;
}

export enum Vehicles {
  Car = 'car',
}
```

But when importing Car it _does_ throw an error:
file2.ts
```typescript
export interface Car {
  color: string;
}
```

file3.ts
```typescript
import { Car } from './file2';

export const car: Car = {
  color: 'red',
}

export enum Vehicles {
  Car = 'car',
}
```
